src/coreclr/jit/morph.cpp

```diff
@@ -11454,7 +11454,10 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 }
                 // ARM64 architecture manual suggests this transformation
                 // for the mod operator.
-                else
+                // However, we do skip this optimization for ARM64 if the second operand
```

Comment:
Can you elaborate why we decide to handle it in lower instead of transforming it in morph?

Reply:
The logically equivalent expressions to the ARM64 instructions that emit this optimization with `csneg` requires a conditional.

```
//     {expr} % {cns}
//     Logically turns into:
//         let a = {expr}
//         if a > 0 then (a & ({cns} - 1)) else -(-a & ({cns} - 1))
//     which then turns into:
//         and   reg1, reg0, #({cns} - 1)
//         negs  reg0, reg0
//         and   reg0, reg0, #({cns} - 1)
//         csneg reg0, reg1, reg0, mi
```

As far as I know, I cannot introduce a conditional without creating new basic blocks. Also, GT_QMARK nodes are not allowed to be created in morph, which in theory could handle the conditional cases.

I figured creating new basic blocks to handle something like this might be overkill for the specific thing we are trying to do, which is to optimize `a % b`. This is why it's done in lowering for ARM64 specifically.

Reply:
We eventually are going to introduce something like GT_SELECT for cmov-like optimizations.
The optimization you're trying to implement is not arm64 specific. on x64 signed mod by Power-of-two should also be implemented with a conditional instruction, see https://godbolt.org/z/1rEenqorW

Reply:
Those look equivalent to me, though the instructions I'm emitting:
```
and     w1, w0, #15
negs    w0, w0
and     w0, w0, #15
csneg   w0, w1, w0, mi
```
are different though.

Reply:
>  Though, I don't know if this PR is the place to do that.

Agree. May be leave a TODO note in the comments about what it.